vault/vault.go

```diff
@@ -0,0 +1,183 @@
+package vault
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/vault/api"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	kubernetesJwtTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	kubernetesAuthUrl      = "auth/kubernetes/login"
+)
+
+var log logr.Logger
+
+func fileExists(name string) (bool, error) {
+	_, err := os.Stat(name)
+	if err == nil {
+		return true, nil
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	}
+	return false, err
+}
+
+func vaultClient() (*api.Client, error) {
+	var client *api.Client
+	var err error
+	var vaultSkipVerify bool = false
+
+	vaultUrl := os.Getenv("VAULT_ADDR")
+	if os.Getenv("VAULT_SKIP_VERIFY") != "" && os.Getenv("VAULT_SKIP_VERIFY") == "true" {
+		vaultSkipVerify = true
+	}
+	if vaultUrl == "" {
+		return client, fmt.Errorf("VAULT_ADDR is not set")
+	}
+
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: vaultSkipVerify},
+	}
+
+	httpClient := &http.Client{Transport: tr}
+
+	// create a vault client
+	client, err = api.NewClient(&api.Config{Address: vaultUrl, HttpClient: httpClient})
+	if err != nil {
+		return client, err
+	}
+
+	return client, err
+}
+
+func renewToken() (float64, error) {
+	var tokenTTL float64
+	client, err := vaultClient()
+	if err != nil {
+		return -1, err
+	}
+
+	r, err := client.Logical().Read("auth/token/lookup-self")
+	if err != nil {
+		return -1, err
+	}
+
+	t, err := r.TokenTTL()
+	if err != nil {
+		return -1, err
+	}
+	tokenTTL = t.Seconds()
+
+	/* FIXME: is this a sensible minimum? */
+	if tokenTTL > 120 {
+		return tokenTTL, nil
+	}
+
+	if ok, err := r.TokenIsRenewable(); !ok {
+		return tokenTTL, err
+	}
+
+	r, err = client.Logical().Write("/auth/token/renew-self", map[string]interface{}{
+		"token": os.Getenv("VAULT_TOKEN"),
+	})
+	if err != nil {
+		return tokenTTL, err
+	}
+
+	tokenTTL = float64(r.Auth.LeaseDuration)
+	log.Info(fmt.Sprintf("Vault token renewed. New TTL is %v seconds", tokenTTL))
+
+	return tokenTTL, nil
+}
+
+func kubeLogin() error {
+	if ok, err := fileExists(kubernetesJwtTokenPath); !ok {
+		return err
+	}
+	if os.Getenv("VAULT_ROLE_ID") == "" {
+		return fmt.Errorf("VAULT_ROLE_ID not defined")
+	}
+	client, err := vaultClient()
+	if err != nil {
+		return err
+	}
+	fd, err := os.Open(kubernetesJwtTokenPath)
+	defer fd.Close()
```

Comment:
Move the defer below the if - if err is not nil then fd will be nil and this will cause a nil pointer reference